File: C++/Binary Search Tree/Validate_BST.cpp
// Validate a Binary Search Tree
// Author: Pavan Kumar Paluri
// Time Complexity: O(N) and Space Complexity: O(N)
// Leetcode Question: https://leetcode.com/problems/validate-binary-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool valid_bst(TreeNode* current, long low, long high)
    {
        if(current==nullptr)
            return true;
        if(current->val <= low||current->val >= high)
            return false;
        return valid_bst(current->left, low, current->val) && valid_bst(current->right, current->val, high);
    }
    bool isValidBST(TreeNode* root) {
        return valid_bst(root, LONG_MIN, LONG_MAX);
    }
};

File: Python/Binary Search Tree/Validate_BST.py
# Validate Binary Search Tree 
# Author: Pavan Kumar Paluri
# Leetcode Question: https://leetcode.com/problems/validate-binary-search-tree/
# Time and Space Complexities: O(N) and O(N) 

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        '''
        ------- INORDER APPROACH -------
        ------- TIME COMPLEXITY: O(N) and SPACE: O(2*N) for recursion stack and node stack -------------
        prev= root.val
        list_nodes = []
        def inorder(root):
            if root is None:
                return root
            else:
                inorder(root.left)
                list_nodes.append(root.val)
                inorder(root.right)
        inorder(root)
        # print(list_nodes)
        for i in range(1, len(list_nodes)):
            if list_nodes[i] <= list_nodes[i-1]:
                return False
        return True
        #return param
        '''
        # Approach - Bound. Use hi and low to ensure the vailidity of BST
        low = -math.inf
        hi = math.inf
        def valid_bst(root, low, hi):
            if root is None:
                return True
            # edge cases: 
            if root.val <= low or root.val >= hi:
                # invalid 
                return False
            return valid_bst(root.left, low, root.val) and valid_bst(root.right, root.val, hi)
        return valid_bst(root, low, hi)
